File: bin/collect_new_list.pl
#!/usr/bin/env perl

use English;
use strict;
use warnings;
use Carp;
use Encode;

use HTML::Parser;
use FeedMaker;


# multiline matching
$/ = "";


sub extract_urls
{
	my $render_js = shift;
	my $url = shift;
	my $item_capture_script = shift;
	my $user_agent = shift;
    my $referer = shift;
	#print "# extract_urls($url, $item_capture_script, $user_agent, $referer)\n";

	my $option = "";
	if ($render_js == 1) {
		$option = "--render-js";
	}
	if ($user_agent ne "") {
		$option .= " --ua '$user_agent'";
	}
    if ($referer ne "") {
        $option = "--referer '$referer'";
    }
	my $cmd = qq(wget.sh $option "$url" | extract_element.py collection | $item_capture_script);
	print "# $cmd\n";
	my $result = `$cmd`;
	if ($CHILD_ERROR != 0) {
		confess "Error: can't execute '$cmd', $ERRNO\n";
		return;
	}

	my @result_list = split /\n/, $result;
	# check the result
	foreach my $item (@result_list) {
		if ($item =~ /^\#/) {
			next;
		}
		my @tuple = split /\t/, $item;
		if (scalar @tuple < 2 or $tuple[0] eq "" or $tuple[1] eq "") {
			confess "Error: can't get the link and title from '$item',";
			return;
		}
	}
	return @result_list;
}


sub compose_url_list
{
	my $render_js = shift;
	my $list_url_list = shift ;
	my $item_capture_script = shift;
	my $user_agent = shift;
    my $referer = shift;
	my $total_list = shift;

	if (not defined $list_url_list) {
		$list_url_list = "";
	} 

	print "# compose_url_list($render_js, $list_url_list, $item_capture_script, $referer, $total_list)\n";

	foreach my $key (keys %$list_url_list) {
		my $value = $list_url_list->{$key};
		if (UNIVERSAL::isa($value, 'ARRAY')) {
			# list_url_list 아래 list_url이 여러 개 존재하는 경우
			foreach my $url (@$value) {
				my $a_url = $url;
				my @url_list = extract_urls($render_js, $a_url, $item_capture_script, $user_agent, $referer);
				push @$total_list, @url_list;
			}
		} else {
            # list_url_list 아래 list_url이 하나 존재하는 경우
			my $a_url = $value;
			my @url_list = extract_urls($render_js, $a_url, $item_capture_script, $user_agent, $referer);
			push @$total_list, @url_list;
		}
	}
}


sub main
{
	#print "# main()\n";

	my @total_list = ();

	# configuration
	my $config = ();
	if (not readConfig(\$config)) {
		confess "Error: can't read configuration!,";
		return -1;
	}

	my $render_js = FeedMaker::getConfigValue($config, 0, 0, ("collection", "render_js"));
	if (defined $render_js and $render_js =~ m!(true|yes)!i) {
		$render_js = 1;
	} else {
		$render_js = 0;
	}

	my $list_url_list = $config->{"collection"}->{"list_url_list"};
	if (defined $list_url_list) {
		print "# list_url_list:\n";
		FeedMaker::printAllHashItems($list_url_list, "  - ");
	}

	my $item_capture_script = FeedMaker::getConfigValue($config, 0, "", ("collection", "item_capture_script"));
	if (not defined $item_capture_script or $item_capture_script eq "") {
		$item_capture_script = "./capture_item_link_title.pl";
        if (not -e $item_capture_script) {
            $item_capture_script = "./capture_item_link_title.py";
        }
	}
	print "# item_capture_script: $item_capture_script\n";
	my ($item_capture_script_program,) = split /\s+/, $item_capture_script;
	if ($item_capture_script_program and not -x $item_capture_script_program) {
		confess "Error: can't execute '$item_capture_script_program',";
		return -1;
	}

	my $user_agent = FeedMaker::getConfigValue($config, 0, "", ("collection", "user_agent"));

    my $referer = FeedMaker::getConfigValue($config, 0, "", ("collection", "referer"));

	# collect items from specified url list
	print "# collecting items from specified url list...\n";
	compose_url_list($render_js, $list_url_list, $item_capture_script, $user_agent, $referer, \@total_list);

	foreach my $item (@total_list) {
		print $item . "\n";
	}

	return 0;
}


main();

File: bin/collect_new_list.py
#!/usr/bin/env python3

import pprint
import re
import sys
import time
import getopt
from collections import OrderedDict
from typing import Dict

import feedmakerutil
from feedmakerutil import Config
from feedmakerutil import die
from logger import Logger


logger = Logger("collect_new_list.py")
do_debug = False


def log(*args):
    if do_debug:
        logger.info(*args)
    else:
        logger.debug(*args)


def extract_urls(url, options):
    log("# extract_urls(%s)" % url)
    
    option_str = feedmakerutil.determine_crawler_options(options)
    whole_cmd = ""
    
    cmd = "crawler.py %s '%s'" % (option_str, url)
    if whole_cmd:
        whole_cmd += " | " + cmd
    else:
        whole_cmd += cmd
    log("# %s" % whole_cmd)
    (result, error) = feedmakerutil.exec_cmd(cmd)
    if error:
        time.sleep(5)
        (result, error) = feedmakerutil.exec_cmd(cmd)
        if error:
            log(whole_cmd + "\n" + str(result) + "\n")
            die("can't get result from crawler script")
                
    cmd = "extract_element.py collection"
    if whole_cmd:
        whole_cmd += " | " + cmd
    else:
        whole_cmd += cmd
    log("# %s" % whole_cmd)
    (result, error) = feedmakerutil.exec_cmd(cmd, result)
    if error: 
        log(whole_cmd + "\n" + str(result) + "\n")
        logger.err(error)
        die("can't get result from extract script")

    cmd = options["item_capture_script"]
    if whole_cmd:
        whole_cmd += " | " + cmd
    else:
        whole_cmd += cmd
    log("# %s" % whole_cmd)
    (result, error) = feedmakerutil.exec_cmd(cmd, result)
    if error:
        log(whole_cmd + "\n" + str(result) + "\n")
        die("can't get result from capture script")

    # check the result
    result_list = []
    for line in result.rstrip().split("\n"):
        line = line.rstrip()
        if re.search(r'^#', line) or re.search(r'^\s*$', line):
            continue
        items = line.split("\t")
        link = items[0]
        title = " ".join(items[1:])
        if not link or not title:
            die("can't get the link and title from '%s'," % link)
        result_list.append((link, title))
    return result_list


def compose_url_list(collection_conf: Dict[str, OrderedDict]):
    log("# compose_url_list()")
    result_list = []

    for list_url in collection_conf["list_url_list"]:
        url_list = extract_urls(list_url, collection_conf)
        result_list.extend(url_list)

    result_list = feedmakerutil.remove_duplicates(result_list)
    return result_list


def main():
    global do_debug
    
    try:
        opts, args = getopt.getopt(sys.argv[1:], "d")
    except getopt.GetoptError as err:
        logger.error(err)
        sys.exit(-1)
    
    for o, a in opts:
        if o == "-d":
            do_debug = True
        
    config = Config()
    collection_conf = config.get_collection_configs()
    log("# collection_conf=%s" % pprint.pformat(collection_conf))

    # collect items from specified url list
    log("# collecting items from specified url list...")
    total_list = compose_url_list(collection_conf)
    for (link, title) in total_list:
        logger.info("%s\t%s" % (link, title))

    return 0


if __name__ == "__main__":
    sys.exit(main())
